File: brainfuck2/bf.ex
defmodule BF.Tape do
  alias BF.Tape, as: Tape

  defstruct data: :array.new(default: 0), pos: 0

  def current(tape) do
    :array.get(tape.pos, tape.data)
  end

  def inc(tape, x) when x == -1 or x == 1 do
    new_data = :array.set(tape.pos, :array.get(tape.pos, tape.data) + x,
                          tape.data)
    %Tape{tape | data: new_data}
  end

  def move(tape, x) when x == -1 or x == 1 do
    new_pos = tape.pos + x
    new_data =
      if new_pos < :array.size(tape.data) do
        tape.data
      else
        :array.set(new_pos, 0, tape.data)
      end
    %Tape{tape | data: new_data, pos: new_pos}
  end
end

defmodule BF do
  alias BF.Tape, as: Tape

  def parse(source) do
    {result, []} = parse(String.split(source, ""), [])
    result
  end

  defp parse([], acc) do
    {Enum.reverse(acc), []}
  end

  defp parse([x | xs], acc) do
    case x do
      "+" -> parse(xs, [{:inc, 1} | acc])
      "-" -> parse(xs, [{:inc, -1} | acc])
      ">" -> parse(xs, [{:move, 1} | acc])
      "<" -> parse(xs, [{:move, -1} | acc])
      "." -> parse(xs, [{:print, nil} | acc])
      "[" ->
        {loop_code, new_xs} = parse(xs, [])
        parse(new_xs, [{:loop, loop_code} | acc])
      "]" -> {Enum.reverse(acc), xs}
      _   -> parse(xs, acc)
    end
  end

  def run([], tape) do
    tape
  end

  def run(program = [op | ops], tape) do
    case op do
      {:inc, x} -> run(ops, Tape.inc(tape, x))
      {:move, x} -> run(ops, Tape.move(tape, x))
      {:print, nil} ->
        :ok = IO.binwrite([Tape.current(tape)])
        :file.sync(:stdout)
        run(ops, tape)
      {:loop, loop_code} ->
        if Tape.current(tape) == 0 do
          run(ops, tape)
        else
          run(program, run(loop_code, tape))
        end
    end
  end
end

defmodule Benchmark do
  def run do
    [filename] = System.argv()
    File.read!(filename)
    |> BF.parse()
    |> BF.run(%BF.Tape{})
  end
end

with {:ok, socket} <- :gen_tcp.connect('localhost', 9001, []) do
   :gen_tcp.send(socket, "Elixir")
   :gen_tcp.close(socket)
end

Benchmark.run()

File: brainfuck2/bf.hs
module Main where

import qualified Data.Array.Base as ArrayBase
import qualified Data.Array.Unboxed as UArray
import qualified Data.ByteString.Char8 as C
import Data.Char (chr)
import System.Environment (getArgs)
import System.IO (hFlush, stdout)
import Network.Simple.TCP

data Op = Inc Int | Move Int | Print | Loop [Op] deriving Show
data Tape = Tape { tapeData :: UArray.UArray Int Int
                 , tapePos :: Int
                 } deriving Show

current :: Tape -> Int
current tape = ArrayBase.unsafeAt (tapeData tape) (tapePos tape)

inc :: Int -> Tape -> Tape
inc delta tape =
    tape { tapeData = newData }
  where
    newData = ArrayBase.unsafeReplace (tapeData tape)
                                      [(tapePos tape, (current tape) + delta)]

move :: Int -> Tape -> Tape
move m tape =
    tape { tapeData = newData, tapePos = newPos }
  where
    curData = tapeData tape
    len = ArrayBase.numElements curData
    newPos = (tapePos tape) + m
    asc = ArrayBase.assocs curData
    newData = if newPos < len
              then curData
              else ArrayBase.unsafeArray (0, newPos)
                                         (asc ++ [(i, 0) | i <- [len..newPos]])

parse :: ([Char], [Op]) -> ([Char], [Op])
parse ([], acc) = ([], reverse acc)
parse (c:cs, acc) =
    case c of
        '+'       -> parse (cs, Inc 1:acc)
        '-'       -> parse (cs, Inc (-1):acc)
        '>'       -> parse (cs, Move 1:acc)
        '<'       -> parse (cs, Move (-1):acc)
        '.'       -> parse (cs, Print:acc)
        '['       -> parse (newCs, Loop loop:acc)
                     where (newCs, loop) = parse (cs, [])
        ']'       -> (cs, reverse acc)
        otherwise -> parse (cs, acc)

run :: [Op] -> Tape -> IO Tape
run [] tape = return tape
run (op:ops) tape = do
    case op of
        Inc d -> run ops $ inc d tape
        Move m -> run ops $ move m tape
        Print -> do
            putStr $ [chr $ current tape]
            hFlush stdout
            run ops tape
        Loop loop -> do
            if current tape == 0
            then run ops tape
            else do
                newTape <- run loop tape
                run (op:ops) newTape

main = do
    connect "localhost" "9001" $ \(socket, _) -> do
      send socket $ C.pack "Haskell"
    [filename] <- getArgs
    source <- readFile filename
    let (_, ops) = parse (source, [])
    run ops (Tape (ArrayBase.unsafeArray (0, 0) [(0, 0)]) 0)
